sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,157 @@
+ *      and in any order.
+ *   - After skipRow() returns `true`, the internal state of `JsonFilter` can be inconsistent,
+ *     so, skipRow() must not be called for the current row anymore without reset().
+ *
+ * @param pushedFilters The pushed down source filters. The filters should refer to
+ *                      the fields of the provided schema.
+ * @param schema The required schema of records from datasource files.
+ */
+class JsonFilters(pushedFilters: Seq[sources.Filter], schema: StructType)
+  extends StructFilters(pushedFilters, schema) {
+
+  /**
+   * Stateful JSON predicate that keeps track of its dependent references in the
+   * current row via `refCount`.
+   *
+   * @param predicate The predicate compiled from pushed down source filters.
+   * @param totalRefs The total amount of all filters references which the predicate
+   *                  compiled from.
+   * @param refCount The current number of predicate references in the row that have
+   *                 been not set yet. When `refCount` reaches zero, the predicate
+   *                 has all dependencies are set, and can be applied to the row.
+   */
+  case class JsonPredicate(predicate: BasePredicate, totalRefs: Int, var refCount: Int) {
+    def reset(): Unit = {
+      refCount = totalRefs
+    }
+  }
+
+  // Predicates compiled from the pushed down filters. The predicates are grouped by their
+  // attributes. The i-th group contains predicates that refer to the i-th field of the given
+  // schema. A predicates can be places to many groups if it has many attributes. For example:
+  //  schema: i INTEGER, s STRING
+  //  filters: IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc"))
+  //  predicates:
+  //    0: Array(IsNotNull("i"), AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+  //    1: Array(AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+  private val predicates: Array[Array[JsonPredicate]] = {
+    val groupedPredicates = Array.fill(schema.length)(Array.empty[JsonPredicate])
+    if (SQLConf.get.jsonFilterPushDown) {
+      val groupedByRefSet = filters
+        // Group filters that have the same set of references. For example:
+        //   IsNotNull("i") -> Set("i"), AlwaysTrue -> Set(),
+        //   And(EqualTo("i", 0), StringStartsWith("s", "abc")) -> Set("i", "s")
+        // By grouping filters we could avoid tracking their state of references in the
+        // current row separately.
+        .groupBy(_.references.toSet)
+        // Combine all filters from the same group by `And` because all filters should
+        // return `true` to do not skip a row. The result is compiled to a predicate.
+        .map { case (refSet, refsFilters) =>
+          (refSet, JsonPredicate(toPredicate(refsFilters), refSet.size, 0))
+        }
+      // Apply predicates w/o references like AlwaysTrue and AlwaysFalse to all fields.
```

Comment:
I think it's better to handle AlwaysTrue and AlwaysFalse in a higher level. We should remove AlwaysTrue. If there is AlwaysFalse, we just return Nil and don't load data at all.

Reply:
1. How often `AlwaysTrue` and `AlwaysFalse` are pushed down in real life? I don't think so often, so, I don't think it make sense to handle the filters in some special way.
2. In the generated Java code (and in final byte code), the checks will be on top:
AlwaysTrue: if (true) bla-bla-bla
AlwaysFalse: if (false) bla-bla-bla
JVM can easily optimize it, and collapse the branches.
3. Now, the code is pretty generic. As you can see, I don't refer to AlwaysTrue and AlwaysFalse. You propose add specific check for the filters somewhere outside of `JsonFilters`, probably, in JacksonParser otherwise I don't know how you are going to prevent loading data at all. This seems premature optimization for me.